### ISSUE TITLE
Unlocking database taking 30-50 seconds to finish

### ISSUE THREAD

**Issue by anshulbajpai** [PhilippC/keepass2android#283]
I am currently having an issue where KP2A is taking at least 30-40 seconds everytime to unlock the database with the master password.

While this happens, I see a message on the screen - `Loading database... (Transforming master key...)`. 

I am using version 1.04b of the app. 
Android version - 8.0.0

**Comment by Karreg** [comment]
Maybe it's related to a too high key transformation iterations? Can you tell us how many iterations are used?

**Comment by anshulbajpai** [comment]
@Karreg Can you tell me where to find it?

**Comment by anshulbajpai** [comment]
I think I have figured it out how to locate the iterations. I think you meant `Key Derivation Function`. 
I recently switched to using `Argon2` from `AES-KDF KDBX 3.1` using another application which I use on mac which is KeyPassXC. The default value there for `Transform Rounds` was 23. 
But when I switched to `Argon2` from KP2A, the default value for `Transform Rounds` is 2. 
I could see the difference of that value between these two applications is quite significant. 
I believe that is why KP2A takes too long to open the database.

Can I ask what does the community think as the safe value for `Transform Rounds` for `Argon2`?

**Comment by anshulbajpai** [comment]
On another note, it is worth mentioning that KeyPassXC doesn't have any lag in opening the database with Argon2 and TansformRounds=23 whereas KP2A has a lag of 30-40 seconds for the same parameters.

Can this not be optimized in KP2A?

**Comment by Karreg** [comment]
The issue here is computing power. This kind of security is here to prevent, are at least harden, brute forcing a database. It requires CPU. And of course, computer and smartphones don't have the same CPU, so it's logical for a smartphone to take more time to open a database. There may be some optimizations that could be done, but it's not the issue here.

On my side, I used the standard Keepass to setup the iteration count (1 second computing). But it was way too long on smartphone, si I lowered it to AES-KDF with 23 million iterations, and opening takes something like 3-10 seconds on my smartphone (snapdragon 821). I prefer a bit more time to open it than a bad security.

You should setup the iteration count to a value that is acceptable for you.

What is your smartphone?

**Comment by anshulbajpai** [comment]
Yea, that's a fair point in favour of the smartphones. I am using a One Plus 3.

As I said before, even I was using AES-KDF with the number of iterations in order or millions and it was was very quick (less than a second) to unlock database both on my phone and mac.

But then I switched to Argon2 (since KeePassXC suggested it is recommended) but the default number of rounds (23 - yea just 23, not in millions) is taking 30 seconds to unlock. 

I definitely know 30 seconds is not acceptable to me but then I wanted to know what is the minimum number of iterations that people are keeping with Argon2, so that I could configure something around that value which is also acceptable to me in terms of unlocking duration.

**Comment by mydexterid** [comment]
Yep.. the current implementation of argon2 (which is a reimplementation of the original C code in C# (https://github.com/PhilippC/keepass2android/blob/master/src/KeePassLib2Android/Cryptography/KeyDerivation/Argon2Kdf.Core.cs)) is really slow on mobile.

Maybe it could be quicker using a native implementation? IDK, could be worth a try.

**Comment by Karreg** [comment]
I see that Argon2 is the next recommendation, but is relatively new in Keepass, so I guess there's room for improvement indeed.
I guess "make it work first, then make it work faster" :)

**Comment by hackel** [comment]
I experimented a bit with this prior to upgrading to kdbx4/argon2.  Even if I set the encryption rounds as high as 100, I'm able to unlock a new database in only 2-3 seconds.  My Pixel is also using a Snapdragon 821.

I just used some random image file as a key with a test database.  Is this the proper way to be testing the speed?  On my laptop (i7), KeepassXC calculates 27 rounds for a 1-second delay.  There's no way my phone should be that fast, right?  In any case, I'm not able to reproduce the 30-50 second unlocking delay.
